tools/src/main/scala/scala/scalanative/codegen/llvm/CodeGen.scala

```diff
@@ -68,12 +68,41 @@ object CodeGen {
       val env = assembly.map(defn => defn.name -> defn).toMap
       val workDir = VirtualDirectory.real(config.workDir)
 
+      /**
+       *    - \C:a\b\c.txt -> a\b\c.txt
+       *    - C:a\b\c.txt -> a\b\c.txt
+       *    - p:\a\b\c.txt -> a\b\c.txt
+       *    - /a/b/c.txt -> a/b/c.txt
+       */
+      def dropPrefix(fileName: String): String = {
+        val driveLetter = s"^[A-Za-z]:".r
+        val supportDriveLetter =
+          build.Platform.isWindows || build.Platform.isCygwin || build.Platform.isMsys
+        // using "/" instead of `File.separator`
+        // we get "/C:a/b/c.txt" even in windows for some reasons.
+        val noDriveLetter =
+          driveLetter.findFirstIn(fileName.stripPrefix("/")) match {
+            case Some(prefix) if supportDriveLetter =>
+              fileName.stripPrefix(prefix)
+            case _ => fileName
+          }
+        noDriveLetter.stripPrefix("/")
+      }
+
+      def sourceDirOf(pos: Position): String = {
+        lazy val empty = "__empty"
+        if (pos == null) empty
+        else pos.dir.getOrElse(empty)
+      }
+
       // Partition into multiple LLVM IR files proportional to number
       // of available processesors. This prevents LLVM from optimizing
       // across IR module boundary unless LTO is turned on.
       def separate(): Future[Seq[Path]] =
         Future
-          .traverse(partitionBy(assembly, procs)(_.name.top).toSeq) {
+          .traverse(
+            partitionBy(assembly, procs)(x => sourceDirOf(x.pos)).toSeq
```

Comment:
Since `sourceDirOf` method is only used her maybe let's pass the `Defn` directly to it and use just a function val here?